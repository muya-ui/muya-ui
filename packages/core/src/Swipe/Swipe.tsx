import React, { useMemo, useRef } from 'react';

import { useForkRef } from '@muya-ui/utils';

import memoForwardRef from '../utils/memoForwardRef';
import useStyles from '../utils/useStyles';
import SwipeContainer from './SwipeContainer';
import SwipePanel from './SwipePanel';
import { ISwipeProps } from './types';
import useInnerSwipe from './useInnerSwipe';

const defaultStyles = {
  panel: '',
};

const Swipe = memoForwardRef<HTMLDivElement, ISwipeProps>((props, ref) => {
  const {
    direction = 'horizontal',
    stepIndex,
    itemIndex,
    equalNum,
    gutter,
    offset,
    duration,
    onStepsChange,
    children,
    onTransitionEnd,
    styles,
    defaultIndex,
    enableDiffChildren,
    transitionStatus: propsTransitionStatus,
    ...otherProps
  } = props;
  const innerStyles = useStyles('swipe', defaultStyles, styles);
  const containerRef = useRef<HTMLDivElement>(null);
  const handleRef = useForkRef<HTMLDivElement>(containerRef, ref);
  const { nodes, panelStyle, transitionStatus, finalDuration } = useInnerSwipe(props, containerRef);

  const innerPanelStyle = useMemo(() => {
    if (innerStyles.panel.style) {
      return {
        ...innerStyles.panel.style,
        ...panelStyle,
      };
    }

    return panelStyle;
  }, [innerStyles.panel.style, panelStyle]);

  return (
    <SwipeContainer $direction={direction} {...otherProps} ref={handleRef}>
      <SwipePanel
        className={innerStyles.panel.className}
        $duration={finalDuration}
        $transition={propsTransitionStatus || transitionStatus}
        onTransitionEnd={onTransitionEnd}
        style={innerPanelStyle}
        $direction={direction}
      >
        {nodes}
      </SwipePanel>
    </SwipeContainer>
  );
});

export default Swipe;
