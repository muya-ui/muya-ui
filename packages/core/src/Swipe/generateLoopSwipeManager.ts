import { ILoopSwipeManager } from './types';

export function getItemOffsets(
  items: ClientRect[],
  prevKey: 'left' | 'top',
  panelStartPoint: number,
) {
  return items.map(item => {
    const itemPrev = item[prevKey];
    const offset = itemPrev - panelStartPoint;
    return offset;
  });
}

/**
 * 根据第一页所在的位置计算出所有在面板可能出现的位置
 * @param firstOffset 第一页所在的 offset ，注意这个第一页指 clone 的第一个节点所在的位置
 * @param pageSize 注意这个 page size 指容器的宽度或者长度
 */
export function calcStepsByPageOneInCloneItems(firstOffset: number, pageSize: number) {
  const steps: number[] = [firstOffset, firstOffset + pageSize];
  let i = 1;
  let offset = firstOffset - pageSize * i;
  while (offset >= 0) {
    steps.unshift(offset);
    i++;
    offset = firstOffset - pageSize * i;
  }
  return steps;
}

/**
 * 根据offset找到数组中的 index
 * @param offset
 * @param offsets 一定是正数数组，并是从小到大排列的，由 calcStepsByPageOneInCloneItems 产生
 */
export function findIndexByOffset(offset: number, offsets: number[]) {
  for (let l = offsets.length, i = l - 1; i >= 0; i--) {
    const currentOffset = offsets[i];
    if (currentOffset <= offset) {
      return i;
    }
  }
  return 0;
}

export function findFirstItemIndexInPageTwo(itemOffsets: number[], containerSize: number) {
  for (let i = 0, l = itemOffsets.length; i < l; i++) {
    const offset = itemOffsets[i];
    if (offset > containerSize) {
      return i;
    }
  }
  return 0;
}

interface IInnerMethods {
  readonly pageOffsets: number[];
  readonly itemOffsets: number[];
  getCurrentIndexByRealIndex(inputIndex: number): number;
  getRealIndexByCurrentIndex(inputIndex: number): number;
  getCurrentItemIndexByRealItemIndex(inputIndex: number): number;
  getRealItemIndexByCurrentItemIndex(inputIndex: number): number;
}

interface ILoopSwipeManagerArgs {
  items: ClientRect[];
  cloneItemRects: ClientRect[];
  containerSize: number; // 宽或高
  direction: 'horizontal' | 'vertical';
  gutter: number;
  stepIndex: number;
  itemIndex: number;
  defaultIndex: number;
}

export function generateDisabledLoopSwipeManager(): ILoopSwipeManager {
  return {
    get index() {
      return 0;
    },
    get itemIndex() {
      return 0;
    },
    get realItemIndex() {
      return 0;
    },
    get size() {
      return 0;
    },
    get itemSize() {
      return 0;
    },
    get realItemSize() {
      return 0;
    },
    get offset() {
      return 0;
    },
    get direction() {
      return 'horizontal' as 'horizontal';
    },
    get locked() {
      return true;
    },
    lock() {},
    unlock() {},
    prevItem() {},
    nextItem() {},
    goToItem() {},
    next() {},
    prev() {},
    goTo() {},
    hasStep() {
      return false;
    },
    hasItem() {
      return false;
    },
    getStepOffset() {
      return 0;
    },
    getItemOffset() {
      return 0;
    },
    updateOffset() {},
  };
}

export default function generateLoopSwipeManager(
  args: ILoopSwipeManagerArgs,
): ILoopSwipeManager & IInnerMethods {
  const {
    items,
    cloneItemRects,
    containerSize, // 宽或高
    direction,
    gutter: innerGutter,
    stepIndex,
    itemIndex,
    defaultIndex = 0,
  } = args;
  const isH = direction === 'horizontal';
  const prevKey = isH ? 'left' : 'top';
  const panelStartPoint = items[0][prevKey];
  const itemOffsets = getItemOffsets(items, prevKey, panelStartPoint);
  const cloneItemOffsets = getItemOffsets(cloneItemRects, prevKey, panelStartPoint);
  // 公共的状态
  let currentOffset = 0;
  let locked = false;

  const stepContainerSize = containerSize + innerGutter;

  // item 的状态
  const itemSize = items.length;
  let currentItemIndex = itemIndex > 0 ? itemIndex : defaultIndex;
  let realItemIndex = currentItemIndex;
  let prevRealItemIndex = currentItemIndex;
  const cloneFirstOffset = cloneItemOffsets[0];
  const realItemOffsets = [...itemOffsets, ...cloneItemOffsets];
  const getCurrentItemIndexByRealItemIndex = (inputIndex: number) => {
    if (inputIndex >= itemOffsets.length) {
      return inputIndex - itemOffsets.length;
    }
    return inputIndex;
  };
  const getRealItemIndexByCurrentItemIndex = (inputIndex: number) => {
    if (inputIndex > 1) {
      return inputIndex;
    }
    const indexA = inputIndex;
    const indexB = inputIndex + itemSize;
    const prevOffset = realItemOffsets[prevRealItemIndex];
    const offsetA = realItemOffsets[indexA];
    const offsetB = realItemOffsets[indexB];

    if (Math.abs(offsetA - prevOffset) < Math.abs(offsetB - prevOffset)) {
      return indexA;
    }
    return indexB;
  };
  if (currentItemIndex && currentItemIndex >= 1) {
    realItemIndex = currentItemIndex;
  } else if (currentItemIndex === 0) {
    realItemIndex = itemSize;
  }
  currentItemIndex = getCurrentItemIndexByRealItemIndex(realItemIndex);
  prevRealItemIndex = realItemIndex;

  const checkLockedByCurrentItemIndex = (inputIndex: number) => {
    const innerRealItemIndex = getRealItemIndexByCurrentItemIndex(inputIndex);
    if (innerRealItemIndex <= 0) {
      locked = true;
    } else if (innerRealItemIndex > itemSize) {
      locked = true;
    }
  };

  // 因为循环的原因，第一页所在的位置实际是会变更的
  let currentIndex = 0;
  let realIndex = currentIndex;
  let prevRealIndex = 0;
  let pageOffsets = calcStepsByPageOneInCloneItems(cloneFirstOffset, stepContainerSize);
  let stepSize = pageOffsets.length - 2;
  if (itemOffsets[defaultIndex] < stepContainerSize) {
    currentOffset = cloneItemOffsets[defaultIndex];
  } else {
    currentOffset = itemOffsets[defaultIndex];
  }
  const getCurrentIndexByRealIndex = (inputIndex: number) => {
    if (inputIndex >= stepSize) {
      return inputIndex - stepSize;
    }
    return inputIndex;
  };
  const getRealIndexByCurrentIndex = (inputIndex: number) => {
    if (inputIndex > 1) {
      return inputIndex;
    }
    const indexA = inputIndex;
    const indexB = inputIndex + stepSize;
    const prevOffset = pageOffsets[prevRealIndex];
    const offsetA = pageOffsets[indexA];
    const offsetB = pageOffsets[indexB];

    if (Math.abs(offsetA - prevOffset) < Math.abs(offsetB - prevOffset)) {
      return indexA;
    }
    return indexB;
  };
  realIndex =
    stepIndex && stepIndex > 0 ? stepIndex : findIndexByOffset(currentOffset, pageOffsets);
  currentIndex = getCurrentIndexByRealIndex(realIndex);
  prevRealIndex = realIndex;

  const checkLockedByCurrentIndex = (inputIndex: number) => {
    const innerRealIndex = getRealIndexByCurrentIndex(inputIndex);
    if (innerRealIndex === 0 || innerRealIndex === pageOffsets.length - 1) {
      locked = true;
    }
  };

  return {
    get index() {
      return currentIndex;
    },
    get itemIndex() {
      return currentItemIndex;
    },
    get realItemIndex() {
      return currentItemIndex;
    },
    get size() {
      return stepSize;
    },
    get itemSize() {
      return itemSize;
    },
    get realItemSize() {
      return itemSize;
    },
    get offset() {
      return currentOffset;
    },
    get direction() {
      return direction;
    },
    get locked() {
      return locked;
    },
    get pageOffsets() {
      return pageOffsets;
    },
    get itemOffsets() {
      return realItemOffsets;
    },
    getCurrentIndexByRealIndex,
    getRealIndexByCurrentIndex,
    getCurrentItemIndexByRealItemIndex,
    getRealItemIndexByCurrentItemIndex,
    lock() {
      locked = true;
    },
    unlock() {
      locked = false;
      const innerRealIndex = getRealIndexByCurrentIndex(currentIndex);
      if (innerRealIndex === 0) {
        const firstPageOffset = pageOffsets[0] + cloneFirstOffset;
        pageOffsets = calcStepsByPageOneInCloneItems(firstPageOffset, stepContainerSize);
        stepSize = pageOffsets.length - 2;
        prevRealIndex = stepSize;
      } else if (innerRealIndex === pageOffsets.length - 1) {
        const baseOffset =
          pageOffsets[pageOffsets.length - 1] - cloneFirstOffset - stepContainerSize;
        const remainOffset = (cloneFirstOffset - baseOffset) % stepContainerSize;
        const firstPageOffset = remainOffset > 0 ? stepContainerSize - remainOffset : 0;
        const firstPageOffsetInCloneRegion = firstPageOffset + cloneFirstOffset;
        pageOffsets = calcStepsByPageOneInCloneItems(
          firstPageOffsetInCloneRegion,
          stepContainerSize,
        );
        stepSize = pageOffsets.length - 2;
        prevRealIndex = 1;
      }

      const innerRealItemIndex = getRealItemIndexByCurrentItemIndex(currentItemIndex);
      if (innerRealItemIndex === 0) {
        prevRealItemIndex = itemSize;
      } else if (innerRealItemIndex >= itemSize) {
        prevRealItemIndex = innerRealItemIndex - itemSize;
      }
    },
    prevItem() {
      if (locked) return;
      prevRealItemIndex = getRealItemIndexByCurrentItemIndex(currentItemIndex);
      currentItemIndex = currentItemIndex - 1;
      if (currentItemIndex < 0) {
        currentItemIndex = itemOffsets.length - 1;
      }
      checkLockedByCurrentItemIndex(currentItemIndex);
    },
    nextItem() {
      if (locked) return;
      prevRealItemIndex = getRealItemIndexByCurrentItemIndex(currentItemIndex);
      currentItemIndex = currentItemIndex + 1;
      if (currentItemIndex >= itemOffsets.length) {
        currentItemIndex = 0;
      }
      checkLockedByCurrentItemIndex(currentItemIndex);
    },
    goToItem(itemIndex) {
      if (locked) return;
      if (this.hasItem(itemIndex)) {
        prevRealItemIndex = getRealItemIndexByCurrentItemIndex(currentItemIndex);
        currentItemIndex = itemIndex;
        checkLockedByCurrentItemIndex(itemIndex);
      }
    },
    next() {
      if (locked) return;
      prevRealIndex = getRealIndexByCurrentIndex(currentIndex);
      currentIndex = currentIndex + 1;
      if (currentIndex >= stepSize) {
        currentIndex = 0;
      }
      checkLockedByCurrentIndex(currentIndex);
    },
    prev() {
      if (locked) return;
      prevRealIndex = getRealIndexByCurrentIndex(currentIndex);
      currentIndex = currentIndex - 1;
      if (currentIndex < 0) {
        currentIndex = stepSize - 1;
      }
      checkLockedByCurrentIndex(currentIndex);
    },
    goTo(stepIndex) {
      if (locked) return;
      if (this.hasStep(stepIndex)) {
        prevRealIndex = getRealIndexByCurrentIndex(currentIndex);
        currentIndex = stepIndex;
        checkLockedByCurrentIndex(currentIndex);
      }
    },
    hasStep(stepIndex) {
      return stepIndex >= 0 && stepIndex < pageOffsets.length;
    },
    hasItem(itemIndex) {
      return itemIndex >= 0 && itemIndex < itemSize;
    },
    getStepOffset(stepIndex) {
      const newRealIndex = getRealIndexByCurrentIndex(stepIndex);
      prevRealIndex = newRealIndex;
      currentIndex = stepIndex;
      checkLockedByCurrentIndex(stepIndex);
      currentOffset = pageOffsets[newRealIndex];
      return currentOffset;
    },
    getItemOffset(inputItemIndex) {
      const newRealItemIndex = getRealItemIndexByCurrentItemIndex(inputItemIndex);
      currentItemIndex = inputItemIndex;
      prevRealItemIndex = newRealItemIndex;
      checkLockedByCurrentItemIndex(currentItemIndex);
      currentOffset = realItemOffsets[newRealItemIndex];
      return currentOffset;
    },
    updateOffset(offset: number) {
      currentOffset = offset;
    },
  };
}
