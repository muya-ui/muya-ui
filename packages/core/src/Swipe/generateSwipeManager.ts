import { ISwipeManager } from './types';

function getStepsArr(
  items: ClientRect[],
  containerSize: number,
  isH: boolean,
  defaultIndex: number = 0,
) {
  const itemSteps: number[] = [];
  const positionSteps: number[] = [];
  const reverseSteps: number[] = [];
  const prevKey = isH ? 'left' : 'top';
  const nextKey = isH ? 'right' : 'bottom';

  const panelStartPoint = items[0][prevKey];
  const panelEndPoint = items[items.length - 1][nextKey];
  let baseline = panelStartPoint + containerSize;
  const panelSize = panelEndPoint - panelStartPoint;
  let defaultStepIndex = 0;

  positionSteps.push(0);
  if (panelSize < containerSize) {
    return { itemSteps, positionSteps, reverseSteps, defaultStepIndex };
  }

  for (let i = 0, l = items.length; i < l; i++) {
    const item = items[i];
    const itemNext = item[nextKey];
    const itemPrev = item[prevKey];
    if (itemNext > baseline) {
      baseline = itemPrev + containerSize;
      const offset = itemPrev - panelStartPoint;
      positionSteps.push(offset);
    }
    if (i === defaultIndex) {
      defaultStepIndex = positionSteps.length - 1;
    }
    itemSteps.push(positionSteps.length - 1);
  }

  const lastPageSize = panelEndPoint - (positionSteps[positionSteps.length - 1] + panelStartPoint);
  if (lastPageSize < containerSize) {
    positionSteps.pop();
    const lastPositionOffset = panelEndPoint - panelStartPoint - containerSize;
    positionSteps.push(lastPositionOffset);
    baseline = panelEndPoint - containerSize;
    reverseSteps.push(lastPositionOffset);
    for (let i = items.length - 1; i >= 0; i--) {
      const item = items[i];
      const itemNext = item[nextKey];
      const itemPrev = item[prevKey];
      if (itemPrev < baseline) {
        baseline = itemNext - containerSize;
        const offset = itemNext - panelStartPoint - containerSize;
        reverseSteps.push(offset);
      }
    }
    reverseSteps.pop();
    reverseSteps.push(0);
    reverseSteps.sort((a, b) => a - b);
  }
  return { itemSteps, positionSteps, reverseSteps, defaultStepIndex };
}

export function getItemOffsets(items: ClientRect[], lastPageOffset: number, isH: boolean) {
  const prevKey = isH ? 'left' : 'top';
  const panelStartPoint = items[0][prevKey];
  const offsets: number[] = [];

  let hasLast = true;
  items.forEach(item => {
    const itemPrev = item[prevKey];
    const offset = itemPrev - panelStartPoint;
    if (offset <= lastPageOffset) {
      offsets.push(offset);
      if (offset === lastPageOffset) {
        hasLast = false;
      }
    }
  });
  if (hasLast) {
    offsets.push(lastPageOffset);
  }

  return offsets;
}

interface ISwipeManagerArgs {
  items: ClientRect[];
  containerSize: number; // 宽或高
  direction: 'horizontal' | 'vertical';
  stepIndex: number;
  itemIndex: number;
  defaultIndex?: number;
}

export function generateDisabledSwipeManager(): ISwipeManager {
  return {
    get index() {
      return 0;
    },
    get itemIndex() {
      return 0;
    },
    get hasNext() {
      return false;
    },
    get hasPrev() {
      return false;
    },
    get hasNextItem() {
      return false;
    },
    get hasPrevItem() {
      return false;
    },
    get size() {
      return 0;
    },
    get itemSize() {
      return 0;
    },
    get offset() {
      return 0;
    },
    get maxOffset() {
      return 0;
    },
    get direction() {
      return 'horizontal' as 'horizontal';
    },
    prevItem() {},
    nextItem() {},
    goToItem() {},
    next() {},
    prev() {},
    goTo() {},
    hasItem() {
      return false;
    },
    hasStep() {
      return false;
    },
    getItemOffset() {
      return 0;
    },
    getItemStep() {
      return 0;
    },
    getStepOffset() {
      return 0;
    },
    updateOffset() {},
  };
}

export default function generateSwipeManager(args: ISwipeManagerArgs): ISwipeManager {
  const { items, containerSize, direction, stepIndex, itemIndex, defaultIndex = 0 } = args;
  const isH = direction === 'horizontal';
  let steps: number[];
  const { itemSteps, positionSteps, reverseSteps, defaultStepIndex } = getStepsArr(
    items,
    containerSize,
    isH,
    defaultIndex,
  );
  const itemOffsets = getItemOffsets(items, positionSteps[positionSteps.length - 1], isH);
  // 公共的状态
  let currentOffset = 0;
  let maxOffset = 0;

  // step 的状态
  steps = positionSteps;
  let currentIndex = stepIndex >= 0 ? stepIndex : defaultStepIndex;
  maxOffset = steps[steps.length - 1];
  let nextIndex = currentIndex + 1;
  let prevIndex = currentIndex - 1;

  // item 的状态
  let currentItemIndex = itemIndex >= 0 ? itemIndex : defaultIndex;
  let prevItemIndex = currentItemIndex - 1;
  let nextItemIndex = currentItemIndex + 1;

  const refreshItemIndex = () => {
    for (let i = 0, l = itemOffsets.length; i < l; i++) {
      const itemOffset = itemOffsets[i];
      if (itemOffset === currentOffset) {
        currentItemIndex = i;
        nextItemIndex = currentItemIndex! + 1;
        prevItemIndex = currentItemIndex! - 1;

        break;
      } else if (itemOffset > currentOffset) {
        nextItemIndex = i;
        prevItemIndex = nextItemIndex - 1;
        currentItemIndex = i;
        break;
      }
    }
  };

  const refreshStepIndex = () => {
    for (let i = 0, l = steps.length; i < l; i++) {
      const stepOffset = steps[i];
      if (stepOffset === currentOffset) {
        currentIndex = i;
        nextIndex = currentIndex + 1;
        prevIndex = currentIndex - 1;
        break;
      } else if (stepOffset > currentOffset) {
        nextIndex = i;
        prevIndex = nextIndex - 1;
        currentIndex = i;
        break;
      }
    }
  };

  const updateSteps = () => {
    if (reverseSteps.length && currentIndex === steps.length - 1) {
      steps = reverseSteps;
    } else if (reverseSteps.length && currentIndex <= 0) {
      steps = positionSteps;
    }
    prevIndex = currentIndex - 1;
    nextIndex = currentIndex + 1;
    currentOffset = steps[currentIndex];
    refreshItemIndex();
  };

  const updateItems = () => {
    prevItemIndex = currentItemIndex - 1;
    nextItemIndex = currentItemIndex + 1;
    currentOffset = itemOffsets[currentItemIndex];
    refreshStepIndex();
  };

  updateSteps();
  return {
    get index() {
      return currentIndex;
    },
    get itemIndex() {
      return currentItemIndex;
    },
    get hasNext() {
      return steps[nextIndex] !== undefined;
    },
    get hasPrev() {
      return steps[prevIndex] !== undefined;
    },
    get hasNextItem() {
      return itemOffsets[nextItemIndex] !== undefined;
    },
    get hasPrevItem() {
      return itemOffsets[prevItemIndex] !== undefined;
    },
    get size() {
      return steps.length;
    },
    get itemSize() {
      return itemOffsets.length;
    },
    get offset() {
      return currentOffset;
    },
    get maxOffset() {
      return maxOffset;
    },
    get direction() {
      return direction;
    },
    prevItem() {
      if (!this.hasPrevItem) {
        return;
      }
      currentItemIndex = prevItemIndex;
      updateItems();
    },
    nextItem() {
      if (!this.hasNextItem) {
        return;
      }
      currentItemIndex = nextItemIndex;
      updateItems();
    },
    goToItem(itemIndex) {
      if (!this.hasItem(itemIndex)) {
        return;
      }
      currentItemIndex = itemIndex;
      updateItems();
    },
    next() {
      if (!this.hasNext) {
        return;
      }
      currentIndex = nextIndex;
      updateSteps();
    },
    prev() {
      if (!this.hasPrev) {
        return;
      }
      currentIndex = prevIndex;
      updateSteps();
    },
    goTo(stepKey) {
      if (!this.hasStep(stepKey)) {
        return;
      }
      currentIndex = stepKey;
      updateSteps();
    },
    hasItem(itemIndex) {
      return itemOffsets[itemIndex] !== undefined;
    },
    hasStep(stepKey) {
      return steps[stepKey] !== undefined;
    },
    getItemOffset(itemIndex) {
      return itemOffsets[itemIndex];
    },
    getItemStep(itemIndex) {
      return itemSteps[itemIndex] || 0;
    },
    getStepOffset(stepKey) {
      return steps[stepKey];
    },

    // 只有在支持滚轮的时候需要处理这个
    updateOffset(offset: number) {
      if (offset > maxOffset || offset < 0) {
        return;
      }
      currentOffset = offset;
      refreshItemIndex();
      refreshStepIndex();
    },
  };
}
