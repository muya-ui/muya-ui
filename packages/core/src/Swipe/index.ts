export { default } from './Swipe';
export { default as Swipe } from './Swipe';
export { default as LoopSwipe } from './LoopSwipe';
export { default as useSwipe } from './useSwipe';
export { default as useLoopSwipe } from './useLoopSwipe';
export * from './types';
