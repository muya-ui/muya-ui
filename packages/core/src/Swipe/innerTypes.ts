// 把一些不要export出去的types放这里
import { ReactNode, RefObject } from 'react';

import { IThemedBaseProps } from '../types';
import { ISwipeBaseProps, ISwipePureProps } from './types';

export type IDirection = Required<Pick<ISwipePureProps, 'direction'>>;
export interface ISwipePanelPureProps extends IThemedBaseProps, IDirection {
  transition: 'on' | 'off';
  duration: number;
}
export type ISwipePanelProps = ISwipePanelPureProps & React.HTMLAttributes<HTMLDivElement>;

type PropsFromBaseProps = Pick<ISwipeBaseProps, 'gutter' | 'equalNum' | 'direction'>;

export interface ISwipeBaseHookState extends PropsFromBaseProps {
  containerRect?: ClientRect;
  children: ReactNode;
  status: 'init' | 'clone_init' | 'done';
  cloneNum: number;
}

export interface ISwipeBaseHookArgs extends PropsFromBaseProps {
  containerRef: RefObject<HTMLDivElement>;
  children: ReactNode;
  /**
   * 是否开启检查子节点变更
   * 开启了以后要注意，传到 Swipe 中的 children 需使用 React.useMemo 包裹
   */
  enableDiffChildren?: boolean;
  duration?: number;
  /**
   * 是否循环
   */
  loop?: boolean;
  onRectChange: (
    itemRects: ClientRect[],
    containerRect: ClientRect,
    cloneItemRects: ClientRect[],
  ) => void;
}

export interface ISwipeCloneResult {
  refs: RefObject<HTMLDivElement>[];
  cloneRefs: RefObject<HTMLDivElement>[];
  nodes: ReactNode;
}
