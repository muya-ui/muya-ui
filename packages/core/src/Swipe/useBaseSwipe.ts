import { Children, useContext, useEffect, useMemo, useRef, useState } from 'react';
import warning from 'warning';

import ImgPoolContext from '../Img/ImgPoolContext';
import useTheme from '../utils/useTheme';
import { ISwipeBaseHookArgs, ISwipeBaseHookState, ISwipeCloneResult } from './innerTypes';
import {
  calculateCloneNum,
  calculateSize,
  cloneChildren,
  getChildStyle,
  rectSizeEqual,
  transformItems,
} from './utils';

export default function useBaseSwipe({
  containerRef,
  direction = 'horizontal',
  equalNum,
  children,
  gutter,
  duration,
  loop = false,
  enableDiffChildren = false,
  onRectChange,
}: ISwipeBaseHookArgs) {
  const propsState: ISwipeBaseHookState = {
    status: 'init',
    children,
    cloneNum: 0,
    equalNum,
    gutter,
    direction,
  };
  const theme = useTheme();
  const finalDuration = duration || theme.transition.spec.duration.slow;
  const [state, setState] = useState<ISwipeBaseHookState>(propsState);
  if (
    state.equalNum !== equalNum ||
    state.gutter !== gutter ||
    state.direction !== direction ||
    Children.count(state.children) !== Children.count(children) ||
    /**
     * 没有默认打开是因为
     * <Swipe>
     *   <div>1</div>
     *   <div>2</div>
     * </Swipe>
     * 这种，其实 children 是变了
     */
    (state.children !== children && enableDiffChildren)
  ) {
    setState(propsState);
  }
  const mergeState = (newState: Partial<ISwipeBaseHookState>) => {
    setState(prevState => ({
      ...prevState,
      ...newState,
    }));
  };
  const isH = state.direction === 'horizontal';
  const childNum = Children.count(children);
  const innerChildren = enableDiffChildren ? state.children : children;
  const { nodes, refs, cloneRefs } = useMemo<ISwipeCloneResult>(() => {
    if (state.status === 'init' || !state.containerRect) {
      return {
        nodes: state.children,
        refs: [],
        cloneRefs: [],
      };
    }

    const childStyle = getChildStyle(isH, state.containerRect, state.equalNum, state.gutter);
    return cloneChildren(innerChildren, childStyle, state.cloneNum);
  }, [
    state.status,
    state.containerRect,
    state.equalNum,
    state.gutter,
    state.cloneNum,
    state.children,
    isH,
    innerChildren,
  ]);

  const uiChange = useRef(true);
  const updateContainer = (newContainerRect: ClientRect) => {
    uiChange.current = true;
    if (loop && !state.equalNum) {
      mergeState({
        status: 'clone_init',
        cloneNum: 0,
        containerRect: newContainerRect,
      });
      return;
    }
    let cloneNum = 0;
    if (loop && state.equalNum && childNum > state.equalNum) {
      cloneNum = state.equalNum * 3;
    }
    mergeState({
      status: 'done',
      cloneNum: cloneNum,
      containerRect: newContainerRect,
    });
  };
  const initClone = () => {
    const itemRects = transformItems(refs);
    const containerSize = isH ? state.containerRect!.width : state.containerRect!.height;
    const panelSize = calculateSize(itemRects, isH);
    let cloneNum = 0;
    // 要能循环要 clone 3倍容器长度的节点来保证往前往后都能移动
    const threeContainerSize = 3 * containerSize;
    if (panelSize < threeContainerSize) {
      const times = Math.floor(threeContainerSize / panelSize);
      // 先计算要 clone 所有的节点多少倍
      const baseCloneNum = times * itemRects.length;
      const remainSize = threeContainerSize - panelSize * times;
      const remainCloneNum = calculateCloneNum(itemRects, isH, remainSize)!;
      cloneNum = baseCloneNum + remainCloneNum;
    } else {
      cloneNum = calculateCloneNum(itemRects, isH, threeContainerSize)!;
    }
    mergeState({
      status: 'done',
      cloneNum: cloneNum,
      containerRect: state.containerRect,
    });
  };
  const initContainer = () => {
    let newContainerRect;
    if (containerRef.current) {
      newContainerRect = containerRef.current.getBoundingClientRect();
    }
    if (!newContainerRect) {
      warning(false, '[Swipe]: get container fail.');
      return;
    }
    updateContainer(newContainerRect);
  };

  const imgPool = useContext(ImgPoolContext);
  const shouldUpdateContainer = async () => {
    let stateContainerRect = state.containerRect;
    let newContainerRect;
    if (containerRef.current) {
      newContainerRect = containerRef.current.getBoundingClientRect();
    }
    if (!newContainerRect) {
      warning(false, '[Swipe]: get container fail when window is resized.');
      return;
    }
    if (rectSizeEqual(stateContainerRect, newContainerRect)) {
      return;
    }
    updateContainer(newContainerRect);
  };
  useEffect(() => {
    if (state.status === 'init') {
      initContainer();
    } else if (state.status === 'clone_init') {
      initClone();
    } else if (state.status === 'done' && uiChange.current) {
      uiChange.current = false;
      const itemRects = transformItems(refs);
      const cloneItemRects = transformItems(cloneRefs);
      onRectChange(itemRects, state.containerRect!, cloneItemRects);
    }
    imgPool.resizeCheckFns.add(shouldUpdateContainer);
    return () => {
      imgPool.resizeCheckFns.delete(shouldUpdateContainer);
    };
  });
  return {
    finalDuration,
    state,
    shouldUpdateContainer,
    nodes,
  };
}
