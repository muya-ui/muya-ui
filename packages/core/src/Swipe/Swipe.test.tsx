import React from 'react';
import renderer from 'react-test-renderer';
import sinon from 'sinon';

import Swipe from './Swipe';
import * as useBaseSwipe from './useBaseSwipe';

let mockHook: any;
beforeAll(() => {
  mockHook = sinon.stub(useBaseSwipe, 'default');
  mockHook.returns({
    nodes: '1111',
    finalDuration: 10,
    state: {
      status: 'done',
      cloneNum: 0,
    },
    shouldUpdateContainer: async () => {},
  });
});
afterAll(() => {
  mockHook.restore();
});

test('测试 Swipe 正常情况', () => {
  const tree = renderer
    .create(
      <Swipe stepIndex={0}>
        <div>AAA</div>
      </Swipe>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 Swipe 垂直方向', () => {
  const tree = renderer
    .create(
      <Swipe stepIndex={0} direction="vertical">
        <div>AAA</div>
      </Swipe>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
