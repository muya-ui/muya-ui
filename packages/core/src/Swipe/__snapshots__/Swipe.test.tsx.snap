// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试 Swipe 垂直方向 1`] = `
.c0 {
  position: relative;
  overflow-y: hidden;
}

.c1 {
  position: relative;
  -webkit-transition: -webkit-transform 10ms cubic-bezier(0.4,0,0.2,1);
  -webkit-transition: transform 10ms cubic-bezier(0.4,0,0.2,1);
  transition: transform 10ms cubic-bezier(0.4,0,0.2,1);
}

<div
  className="c0"
>
  <div
    className="c1 muya-swipe-panel"
    onTransitionEnd={[Function]}
    style={
      Object {
        "transform": "translate(0, 0px)",
      }
    }
  >
    1111
  </div>
</div>
`;

exports[`测试 Swipe 正常情况 1`] = `
.c0 {
  position: relative;
  overflow-x: hidden;
}

.c1 {
  position: relative;
  -webkit-transition: -webkit-transform 10ms cubic-bezier(0.4,0,0.2,1);
  -webkit-transition: transform 10ms cubic-bezier(0.4,0,0.2,1);
  transition: transform 10ms cubic-bezier(0.4,0,0.2,1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100000px;
  height: 100%;
}

<div
  className="c0"
>
  <div
    className="c1 muya-swipe-panel"
    onTransitionEnd={[Function]}
    style={
      Object {
        "transform": "translate(0px, 0)",
      }
    }
  >
    1111
  </div>
</div>
`;
